src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java

```diff
@@ -668,4 +674,39 @@ public Void run() {
             });
         }
     }
+
+    @Override
+    public void afterIndexShardStarted(IndexShard indexShard) {
+        final ShardId shardId = indexShard.shardId();
+        final Index index = shardId.getIndex();
+
+        // Check if this is a security index shard
+        if (securityIndex.equals(index.getName())) {
+            // Only trigger on primary shard to avoid multiple reloads
+            if (indexShard.routingEntry() != null && indexShard.routingEntry().primary()) {
+                final String threadName = Thread.currentThread().getName();
+
+                try {
+                    setCurrentThreadName("[" + ThreadPool.Names.GENERIC + "]");
+                    if (isSecurityIndexRestoredFromSnapshot(index)) {
+                        LOGGER.info("Security index primary shard {} started - " + "config reloading for snapshot restore", shardId);
+                        reloadConfiguration(CType.values());
+                    }
+                } finally {
+                    setCurrentThreadName(threadName);
+                }
+            }
+        }
+    }
+
+    private boolean isSecurityIndexRestoredFromSnapshot(Index index) {
```

Comment:
Can we move this into SnapshotRestoreHelper and keep `setCurrentThreadName` private?